src/utils/bindtwostepcarettoattribute.js

```diff
@@ -49,93 +108,425 @@ export default function bindTwoStepCaretToAttribute( view, model, emitter, attri
+		//
+		// 		<paragraph><$text attribute="1">foo{}</$text><$text attribute="2">bar</$text></paragraph>
+		//
+		// but DON'T when already in between of them (no attribute selection):
+		//
+		// 		<paragraph><$text attribute="1">foo</$text>{}<$text attribute="2">bar</$text></paragraph>
+		//
+		if ( isBetweenDifferentValues( position, attribute ) && this._hasAttribute ) {
+			this._preventCaretMovement( data );
+			this._removeSelectionAttribute();
+		} else {
+			// ENGAGE 2-SCM when entering an attribute:
+			//
+			// 		<paragraph>foo{}<$text attribute>bar</$text>baz</paragraph>
+			//
+			if ( isAtStartBoundary( position, attribute ) ) {
+				this._preventCaretMovement( data );
+				this._overrideGravity();
 
-		// Moving right.
-		if ( arrowRightPressed ) {
-			// If gravity is already overridden then do nothing.
-			// It means that we already enter `foo<a>{}bar</a>biz` or left `foo<a>bar</a>{}biz` text with attribute
-			// and gravity will be restored just after caret movement.
-			if ( modelSelection.isGravityOverridden ) {
 				return;
 			}
 
-			// If caret sticks to the bound of Text with attribute it means that we are going to
-			// enter `foo{}<a>bar</a>biz` or leave `foo<a>bar{}</a>biz` the text with attribute.
-			if ( isAtAttributeBoundary( position.nodeAfter, position.nodeBefore, attribute ) ) {
-				// So we need to prevent caret from being moved.
-				data.preventDefault();
-				// And override default selection gravity.
-				model.change( writer => writer.overrideSelectionGravity() );
+			// ENGAGE 2-SCM when leaving an attribute:
+			//
+			//		<paragraph>foo<$text attribute>bar{}</$text>baz</paragraph>
+			//
+			if ( isAtEndBoundary( position, attribute ) && this._hasAttribute ) {
+				this._preventCaretMovement( data );
+				this._overrideGravity();
+			}
+		}
+	}
+
+	/**
+	 * Updates the document selection and the view according to the two–step caret movement state
+	 * when moving **backwards**. Executed upon `keypress` in the {@link module:engine/view/view~View}.
+	 *
+	 * @param {module:engine/model/position~Position} position The model position at the moment of the key press.
+	 * @param {module:engine/view/observer/domeventdata~DomEventData} data Data of the key press.
+	 */
+	handleBackwardMovement( position, data ) {
+		const attribute = this.attribute;
+
+		// When the gravity is already overridden...
+		if ( this._isGravityOverridden ) {
+			// ENGAGE 2-SCM & REMOVE SELECTION ATTRIBUTE
+			// when about to leave one attribute value and enter another:
+			//
+			// 		<paragraph><$text attribute="1">foo</$text><$text attribute="2">{}bar</$text></paragraph>
+			//
+			// but DON'T when already in between of them (no attribute selection):
+			//
+			// 		<paragraph><$text attribute="1">foo</$text>{}<$text attribute="2">bar</$text></paragraph>
+			//
+			if ( isBetweenDifferentValues( position, attribute ) && this._hasAttribute ) {
+				this._preventCaretMovement( data );
+				this._restoreGravity();
```

Comment:
I wonder whether the selection should be restored in case. I guess there's a difference when this would be your model:

```html
<paragraph><$text attribute="1" bold="true">foo</$text><$text attribute="2">{}bar</$text></paragraph>
```

Do we want the bold or not.

Reply:
I don't say that current solution is incorrect, I just wonder.

Reply:
In this case

```
<$text attribute="1" bold="true">foo</$text><$text attribute="2">{}bar</$text>
```

<kbd>←</kbd>: the selection has `bold="true"` only (2-SCM active for `attribute`). It's a typing between two attribute boundaries of a different value case.

<kbd>←</kbd>: the selection has `attribute="1" bold="true"`

If the selection lost `bold="true"` in the first <kbd>←</kbd> step I'd personally find it confusing. Is that what you wanted to discuss?

Reply:
I wonder if the selection should have bold after the first left-arrow press.

Reply:
But this is nothing critical.